libraries/protocol/include/scorum/protocol/betting/market.hpp

```diff
@@ -2,16 +2,188 @@
 #include <scorum/protocol/betting/wincase.hpp>
 #include <scorum/protocol/betting/wincase_comparison.hpp>
 
+#include <boost/fusion/container.hpp>
+#include <boost/fusion/algorithm/iteration/for_each.hpp>
+#include <boost/fusion/include/for_each.hpp>
+#include <boost/container/flat_set.hpp>
+
+#include <scorum/utils/static_variant_serialization.hpp>
+
 namespace scorum {
 namespace protocol {
 namespace betting {
-struct market_type
+
+/// This check if strict_wincase_pair_type has valid market kind
+template <typename WincasePair> constexpr bool check_market_kind(market_kind kind, WincasePair&& w)
+{
+    // Only one check for left_wincase is necessary because strict_wincase_pair_type type
+    // guarantees that left_wincase and right_wincase have the same market kind
+    return kind == std::decay_t<decltype(w)>::left_wincase::kind_v;
+}
+
+template <typename WincasePair, typename... WincasePairs>
+constexpr bool check_market_kind(market_kind kind, WincasePair&& first, WincasePairs&&... ws)
+{
+    return check_market_kind(kind, std::forward<WincasePair>(first))
+        && check_market_kind(kind, std::forward<WincasePairs>(ws)...);
+}
+
+using wincase_pairs_type = fc::flat_set<wincase_pair>;
+
+template <class Implementation, market_kind Kind, typename... Wincases> struct base_market_type
+{
+    static_assert(check_market_kind(Kind, Wincases{}...), "All wincases should have same market kind");
+
+    static constexpr market_kind kind = Kind;
+
+    virtual ~base_market_type()
+    {
+    }
+
+    wincase_pairs_type create_wincase_pairs() const
+    {
+        wincase_pairs_type result;
+        boost::fusion::set<Wincases...> wincase_ps;
+        boost::fusion::for_each(wincase_ps, [&](auto wp) {
+            auto wl = dynamic_cast<const Implementation*>(this)->create_wincase(decltype(wp){});
+            result.emplace(std::make_pair(wincase_type(wl), wincase_type(wl.create_opposite())));
```

Comment:
No need to use std::mare_pair here cuz set's emplace method already creates this pair inside the set in-place